src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,377 @@
+            // an insertion, deletion, or both.
+            var indexToEditKinds = new Dictionary<int, SemanticTokenEditKind>();
+
+            foreach (var edit in edits)
+            {
+                // We only care about EditKind.Insert and EditKind.Delete, since they encompass all
+                // changes to the document. All other EditKinds are ignored.
+                switch (edit.Kind)
+                {
+                    case EditKind.Insert:
+                        indexToEditKinds.TryGetValue(edit.NewIndex, out var editKindWithoutInsert);
+                        indexToEditKinds[edit.NewIndex] = editKindWithoutInsert |= SemanticTokenEditKind.Insert;
+                        break;
+                    case EditKind.Delete:
+                        indexToEditKinds.TryGetValue(edit.OldIndex, out var editKindWithoutDelete);
+                        indexToEditKinds[edit.OldIndex] = editKindWithoutDelete |= SemanticTokenEditKind.Delete;
+                        break;
+                    default:
+                        break;
+                }
+            }
+
+            return CombineEditsIfPossible(newGroupedSemanticTokens, indexToEditKinds);
+        }
+
+        private static SemanticTokensEdit[] CombineEditsIfPossible(
+            SemanticToken[] newGroupedSemanticTokens,
+            Dictionary<int, SemanticTokenEditKind> indexToEditKinds)
+        {
+            // This method combines the edits into the minimal possible edits.
+            // For example, if an index contains both an insertion and deletion, we can combine the two
+            // edits into one.
+            // We can also combine edits if we have consecutive edits of certain types:
+            // Delete->Delete, Insert->Insert, Update->Update, Update->Insert, and Update->Delete.
+            // Note for the Update->Insert and Update->Delete cases, any further edits we combine can
+            // only be an Insert or Delete, respectively.
+
+            using var _ = ArrayBuilder<LSP.SemanticTokensEdit>.GetInstance(out var semanticTokensEdits);
+
+            // We sort the edit kinds by index since we need to know what kind of edits surround
+            // a given index in order to potentially combine them into one edit.
+            var editIndexes = indexToEditKinds.Keys.ToArray();
+            Array.Sort(editIndexes);
+
+            for (var i = 0; i < editIndexes.Length; i++)
+            {
+                var initialEditIndex = editIndexes[i];
+                var initialEditKind = indexToEditKinds[initialEditIndex];
+
+                if (initialEditKind.HasFlag(SemanticTokenEditKind.Insert) && initialEditKind.HasFlag(SemanticTokenEditKind.Delete))
+                {
+                    i = AddInsertionDeletionEdit(
+                        newGroupedSemanticTokens, indexToEditKinds, semanticTokensEdits, editIndexes, i, initialEditIndex);
```

Comment:
so initialEditIndex is only passed in for two reasons:

1. to get the grouped semantic token.
2. to determine the start position of the final edit.  

Consider instead passing in those separate things as they will be easier to name and likely clearer.